tests/testthat/test-slice-chop.R

```diff
@@ -168,6 +168,33 @@ test_that("vec_chop() falls back to `[` for shaped objects with no proxy when in
   expect_equal(result, x)
 })
 
+test_that("vec_chop() with data frame proxies always uses the proxy's length info", {
```

Comment:
Just a note that you can also use the `vctrs_proxy` class from helper-s3.R for this kind of tests, see `local_proxy()`. This is a vector class implemented on top of an environment.

Reply:
I just realised that there's a tension between objects like `vctrs_proxy` and S4-style records. The former work better with names pulled from the proxy, whereas the latter works out of the box with names pulled from the input. I think that explains why we had different instincts about where to pull names in our call yesterday.

Reply:
> The former work better with names pulled from the proxy

Oh, I see what you mean. For `vctrs_proxy`, we have no idea how to extract names from the input. We need the developer to give us a proxy with a standardized place for names to live, so we can slice them appropriately. Interesting.

That does suggest that names should always exist on the proxy, and should be pulled from the proxy. We should probably look at updating `vec_slice_impl()` to do this consistently.